contracts/governance/Vesting/fouryear/FourYearVestingLogic.sol

```diff
@@ -0,0 +1,297 @@
+pragma solidity ^0.5.17;
+pragma experimental ABIEncoderV2;
+
+import "./IFourYearVesting.sol";
+import "../../ApprovalReceiver.sol";
+import "./FourYearVestingStorage.sol";
+import "../../../openzeppelin/SafeMath.sol";
+
+/**
+ * @title Four Year Vesting Logic contract.
+ * @notice Staking, delegating and withdrawal functionality.
+ * @dev Deployed by FourYearVestingFactory contract.
+ * */
+contract FourYearVestingLogic is IFourYearVesting, FourYearVestingStorage, ApprovalReceiver {
+	using SafeMath for uint256;
+
+	/* Events */
+	event TokensStaked(address indexed caller, uint256 amount);
+	event VotesDelegated(address indexed caller, address delegatee);
+	event TokensWithdrawn(address indexed caller, address receiver);
+	event DividendsCollected(address indexed caller, address loanPoolToken, address receiver, uint32 maxCheckpoints);
+	event MigratedToNewStakingContract(address indexed caller, address newStakingContract);
+
+	/* Modifiers */
+	/**
+	 * @dev Throws if called by any account other than the token owner or the contract owner.
+	 */
+	modifier onlyOwners() {
+		require(msg.sender == tokenOwner || isOwner(), "unauthorized");
+		_;
+	}
+
+	/**
+	 * @dev Throws if called by any account other than the token owner.
+	 */
+	modifier onlyTokenOwner() {
+		require(msg.sender == tokenOwner, "unauthorized");
+		_;
+	}
+
+	/* Functions */
+
+	/**
+	 * @notice Sets the max interval.
+	 * @param _interval Max interval for which tokens scheduled shall be staked.
+	 * */
+	function setMaxInterval(uint256 _interval) external onlyOwner {
+		maxInterval = _interval;
```

Comment:
Not mandatory, but makes sense to check: _interval % FOUR_WEEKS == 0.

Reply:
Done